client-node-tests/src/integration.test.ts

```diff
@@ -619,6 +619,22 @@ suite('Client integration', () => {
 		middleware.provideCallHierarchyOutgingCalls = undefined;
 		assert.strictEqual(middlewareCalled, true);
 	});
+
+	suite('Will Create Files', () => {
+		test('Static Registration', async () => {
+			const files: lsclient.FileCreate[] = [
+				{ uri: '???' }
+			];
+			const feature = client.getFeature(lsclient.WillCreateFilesRequest.method);
+			isDefined(feature);
+			feature.fillClientCapabilities(???);
+
+			// Trigger a VS Code file creation event
+
+			// Ensure it triggered a well-known file edit
+		});
```

Comment:
@dbaeumer I'm trying to understand how best to write a test for this. I think I need to:

1. Cause VS Code to call the onDidCreateFiles event (though if this only fires for its UI events, I'm not sure how to that programatically)
2. Either verify that the edit returned by the server was applied by VS Code, or at least that the `waitUntil` method was called on the event and passed the expected edit

Do you have any suggestions on the best way to do these? I couldn't find any similar tests to copy (I'd been using `WillSaveTextDocumentWaitUntil` mostly as a template, though I can't find tests for it - are they named something different?).

Thanks!

Reply:
@DanTup I actually assume in the integration tests that VS Code behaves correctly (e.g. if a user requests code complete they call the code complete handler). So what I usually do in the integration tests is to get the feature and then the provider and I then call the methods directly on the provider. See https://github.com/microsoft/vscode-languageserver-node/blob/master/client-node-tests/src/integration.test.ts#L663

You could do the same. It also allows you to test the middleware.